### ISSUE TITLE
v4 opt:region var default value not resolved on deploy

### ISSUE THREAD

**Issue by bwabt** [serverless/serverless#12748]
### Issue description

It appears an old issue has resurfaced, but for v4.
https://github.com/serverless/serverless/issues/6793

Deployment fails when using a dynamic option value syntax (`${opt:region, 'us-east-1'}`) for provider.region.
We are using this syntax elsewhere in our serverless config without issue.

Work-arounds:
- Reverting to serverless v3 resolves the issue.
- Using a static value (e.g. us-east-1) works.

Error:

> Failed to construct URL with https://cloudformation.${opt:region, 'us-east-1'}.amazonaws.com TypeError: Invalid URL
    at new URL (node:internal/url:797:36)
    at getEndpointUrl (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:37129)
    at evaluateEndpointRule (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:38000)
    at evaluateRules (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:38767)
    at evaluateTreeRule (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:38536)
    at evaluateRules (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:38964)
    at evaluateTreeRule (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:38536)
    at evaluateRules (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:38964)
    at resolveEndpoint (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:76:39994)
    at Object.defaultEndpointResolver [as endpointProvider] (file:///home/user/.serverless/releases/4.1.21/package/dist/sf-core.js:155:406674) {
  code: 'ERR_INVALID_URL',
  input: "https://cloudformation.${opt:region, 'us-east-1'}.amazonaws.com"

## Service Config
```
service: myservice
custom:
  1stStack: 1st
  2ndStack: 2nd-${opt:stage, self:provider.stage}
  3rdStack: 3rd-${opt:stage, self:provider.stage}
  gitCommitTracker:
    location: ./version
provider:
  name: aws
  runtime: nodejs20.x
  stage: dev
  region: ${opt:region, 'us-east-1'}
  stackName: ${self:service}-${opt:stage, self:provider.stage}
  deploymentBucket:
    name: ${cf:${self:custom.1stStack}.DeploymentBucket}
  profile: ...
  vpc:
    securityGroupIds:
      - ${cf:${self:custom.1stStack}.LambdaSecurityGroup}
    subnetIds:
      - ${cf:${self:custom.1stStack}.PrivateSubnet3}
      - ${cf:${self:custom.1stStack}.PrivateSubnet4}
  iam:
    ...
  environment:
    ...
  versionFunctions: false
  tags:
    ...
plugins:
  - serverless-step-functions
  - serverless-git-commit-tracker
package:
  individually: true
  patterns:
    - "!./**"
    - config.js
    - utils.js
    - "./version"
layers:
  ...
functions:
  ...
stepFunctions:
  ...
resources:
  ...
```

**Comment by james-cohen** [comment]
This issue also occurs when using SSM to resolve environment variables

```
provider:
  name: aws
  profile: ***
  deploymentMethod: direct
  versionFunctions: false
  runtime: nodejs20.x
  memorySize: 1536
  region: ${opt:region, 'eu-west-2'}
  stage: ${opt:stage}
  environment:
    RDS_PASSWORD: ${ssm:RDS_LAMBDA_PASSWORD}
```

```
⠏ InitializingFailed to construct URL with https://ssm.${opt:region, 'eu-west-2'}.amazonaws.com TypeError: Invalid URL
    at new URL (node:internal/url:804:36)
    at getEndpointUrl (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:37129)
    at evaluateEndpointRule (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:38000)
    at evaluateRules (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:38767)
    at evaluateTreeRule (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:38536)
    at evaluateRules (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:38964)
    at evaluateTreeRule (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:38536)
    at evaluateRules (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:38964)
    at resolveEndpoint (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:76:39994)
    at Object.defaultEndpointResolver [as endpointProvider] (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:144:5099) {
  code: 'ERR_INVALID_URL',
  input: "https://ssm.${opt:region, 'eu-west-2'}.amazonaws.com"
}

✖ Failed to resolve variable 'RDS_LAMBDA_PASSWORD' with resolver 'ssm' and provider 'aws': TypeError: Invalid URL
Error: Failed to resolve variable 'RDS_LAMBDA_PASSWORD' with resolver 'ssm' and provider 'aws': TypeError: Invalid URL
    at #resolveKey (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:942:43321)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async _ResolverManager.resolve (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:942:48156)
    at async #handlePlaceholderNode (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:942:47080)
    at async file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:942:45215
    at async processNodeAndHandleCompletion (file:///Users/james/.serverless/releases/4.1.22/package/dist/sf-core.js:942:38814)
```

**Comment by czubocha** [comment]
Hi @bwabt @james-cohen, we’ll include the fix in the upcoming release.

**Comment by czubocha** [comment]
Hey @bwabt @james-cohen the fix has been released in [v4.1.23](https://github.com/serverless/serverless/releases/tag/v4.1.23). Feel free to test it out.

**Comment by james-cohen** [comment]
Works great, thank you
